src/main/java/org/truffleruby/core/rope/RopeNodes.java

```diff
@@ -1434,6 +1444,69 @@ public int executeHashNotCalculated(Rope rope) {
         }
     }
 
+    @NodeChildren({
+            @NodeChild(type = RubyNode.class, value = "encoding"),
+            @NodeChild(type = RubyNode.class, value = "bytes"),
+            @NodeChild(type = RubyNode.class, value = "start"),
+            @NodeChild(type = RubyNode.class, value = "end")
+    })
+    public abstract static class EncodingLengthNode extends RubyNode {
+
+        public static EncodingLengthNode create() {
+            return RopeNodesFactory.EncodingLengthNodeGen.create(null, null, null, null);
+        }
+
+        public abstract int executeLength(Encoding encoding, byte[] bytes, int start, int end);
+
+        @Specialization(guards = "encoding.isUTF8()")
+        int utf8Length(Encoding encoding, byte[] bytes, int start, int end) {
+            return UTF8Encoding.class.cast(encoding).length(bytes, start, end);
+        }
+
+        @TruffleBoundary
+        @Specialization(guards = "!encoding.isUTF8()")
+        int genericLength(Encoding encoding, byte[] bytes, int start, int end) {
+            return encoding.length(bytes, start, end);
+        }
+
+    }
+
+    @NodeChildren({
+            @NodeChild(type = RubyNode.class, value = "encoding"),
+            @NodeChild(type = RubyNode.class, value = "bytes"),
+            @NodeChild(type = RubyNode.class, value = "start"),
+            @NodeChild(type = RubyNode.class, value = "end")
+    })
+    public abstract static class PreciseLengthNode extends RubyNode {
+
+        public static PreciseLengthNode create() {
+            return RopeNodesFactory.PreciseLengthNodeGen.create(null, null, null, null);
+        }
+
+        public abstract int executeLength(Encoding encoding, byte[] bytes, int start, int end);
+
+        @Specialization
+        int preciseLength(Encoding encoding, byte[] bytes, int start, int end,
+                          @Cached("create()") EncodingLengthNode encodingLengthNode,
+                          @Cached("create()") BranchProfile startTooLargeProfile,
+                          @Cached("create()") BranchProfile characterTooLargeProfile) {
+            if (start >= end) {
+                startTooLargeProfile.enter();
+                return -1 - (1);
+            }
+
+            int n = encodingLengthNode.executeLength(encoding, bytes, start, end);
+
+            if (n > end - start) {
+                characterTooLargeProfile.enter();
+                return StringSupport.MBCLEN_NEEDMORE(n - (end - start));
+            }
```

Comment:
Shouldn't the caller check those conditions?
It's OK to replace existing code, I'm just asking in general.

Reply:
It's called in quite a few places. It represents a single call in MRI (carried over to `StringSupport`), so a lot of the ported String code expects to have that call available. Besides cutting down on duplication, I think it makes the ported code easier to follow.

The caller's job is determining whether to use `PreciseLengthNode` or `EncodingLengthNode`.